ceno_zkvm/src/scheme/prover.rs

```diff
@@ -111,90 +110,53 @@ impl<
         }
         exit_span!(span);
 
-        // keep track of circuit name to index mapping
-        let circuit_name_index_mapping = self
-            .pk
-            .circuit_pks
-            .keys()
-            .enumerate()
-            .map(|(k, v)| (v, k))
-            .collect::<BTreeMap<_, _>>();
         // only keep track of circuits that have non-zero instances
-        let mut num_instances = Vec::with_capacity(self.pk.circuit_pks.len());
-        let mut num_instances_with_rotation = Vec::with_capacity(self.pk.circuit_pks.len());
-        let mut circuit_name_num_instances_mapping = BTreeMap::new();
-        for (index, (circuit_name, ProvingKey { vk, .. })) in self.pk.circuit_pks.iter().enumerate()
-        {
-            // skip omc init on >1 shard
-            if !shard_ctx.is_first_shard() && vk.get_cs().with_omc_init_only() {
+
+        for chip_input in witnesses.iter_sorted() {
+            let pk = self
+                .pk
+                .circuit_pks
+                .get(&chip_input.name)
+                .ok_or(ZKVMError::VKNotFound(
+                    format!("proving key for circuit {} not found", chip_input.name).into(),
+                ))?;
+
+            assert!(!pk.get_cs().is_static_circuit());
+            // include omc init tables iff it's in first shard
+            if !shard_ctx.is_first_shard() && pk.get_cs().with_omc_init_only() {
                 continue;
             }
+
             // num_instance from witness might include rotation
-            if let Some(num_instance) = witnesses
+            let num_instances = chip_input
                 .num_instances
-                .get(circuit_name)
-                .cloned()
-                .and_then(|num_instance| {
-                    if num_instance.iter().sum::<usize>() > 0 {
-                        Some(num_instance)
-                    } else {
-                        None
-                    }
-                })
-                .or_else(|| {
-                    vk.get_cs().is_static_circuit().then(|| {
```

Comment:
i am checking how new logic work after removing this. Because there are some circuit which only has fixed commitment and they will not present in witness rmm

Reply:
Add commit to remove `is_static_circuit` because now EVERY circuit got 1 structural_witness as selector, thus this mechanism does not needed